week-4/learning-outcomes.md

```diff
@@ -0,0 +1,50 @@
+# Week 4 Learning Outcomes
+
+## Node
+
+### Servers
+
+- [ ] **We can create a web server that responds to http requests**
+- [ ] We can conditionally set ports for our server based on the runtime environment
+- [ ] We can read secrets from a .env file
+- [ ] We can prevent sensitive data from being pushed to GitHub
+
+### GET requests
+
+- [ ] We can route requests to the correct handler function(s)
+- [ ] We can serve different types of files to the client
+- [ ] We can read information sent in a URL query string
+
+### POST requests
+
+- [ ] We can handle data received from a POST request
+- [ ] We can ‘chunk’ data sent from the client side on the server side
+- [ ] We can give descriptive names to HTML form input fields so that it is easy to access data in the POST request object
+- [ ] We can parse a request object
```

Comment:
Does this mean parse the request body? I'm guessing you don't want to parse the whole request object?

Reply:
Maybe

- We can parse JSON sent in an HTTP request
- We can stringify JSON and send it via the HTTP response

Reply:
I don't think we should specify JSON, since forms submit urlencoded data by default (or if we do we should specify both and make it clear when they would use each)

Reply:
@oliverjam Ok, in that case, how about the following:

- We can parse JSON-encoded data sent in a **fetch request**
- We can stringify JSON and send it via the HTTP response

I think specifying that it's a fetch request avoids the confusion caused by my previous phrasing. Fetch requests send JSON-encoded data.

Surely we can get students to make both types of request (JSON-encoded fetch and default urlencoded form submission) in this week's project to build a blogging platform? i.e. when you publish a new post it makes sense for a full page reload to occur, so you would use the default form submission, but when you comment on an existing post (stretch goal), it's nice when just the DOM updates without a full page reload (so you'd use fetch).

Reply:
Yeah that's not a bad idea. I just worry that it might get confusing for people—I think there's already a bit of difficulty understanding the difference between browser and server. I was going to suggest that we say "no client-side JS at all" for this project. We can leave the client-side requests for later weeks

Reply:
They'll still have to do the whole body-parsing streams song and dance, it's just the last step that changes from `JSON.parse(data)` to `querystring.parse(data)` 🤷‍♂. Probably not enough of a difference to worry about

Reply:
nice, that sounds like a great way to do it to me! Thanks @arrested-developer @oliverjam for the comments. So, this learning outcome will just be:

- We can parse data sent via a default HTML form submission

And we can make it clear in the project description that they shouldn't be using JavaScript to send the data for this week's project. Then in a later week we can introduce sending/parsing fetch request data.

I'll make another commit to update this 👍